### PR TITLE
chore(clerk-js): Improve accessibility in UserButton and OrganizationSwitcher

### PR DESCRIPTION
**Pull request by panteliselef**
## Description

This is an effort to improve the underlying semantics of UserButton and OrganizationSwitcher and provide a better accessibility experience. 

In order to come up with this changes i followed the implementation of similar components created with 
- [React aria spectrum](https://react-spectrum.adobe.com/react-aria/)
- Shadcn & Radix
- Vercel's team switcher and user button

Tested with 
- Lighthouse
- [WAVE](https://wave.webaim.org/) 

## Checklist

- [ ] `npm test` runs as expected.
- [ ] `npm run build` runs as expected.
- [ ] (If applicable) [JSDoc comments](https://jsdoc.app/about-getting-started.html) have been added or updated for any package exports
- [ ] (If applicable) [Documentation](https://github.com/clerkinc/clerk-docs) has been updated

## Type of change

- [ ] 🐛 Bug fix
- [ ] 🌟 New feature
- [ ] 🔨 Breaking change
- [ ] 📖 Refactoring / dependency upgrade / documentation
- [x] other:

## Packages affected

- [ ] `@clerk/clerk-js`
- [ ] `@clerk/clerk-react`
- [ ] `@clerk/nextjs`
- [ ] `@clerk/remix`
- [ ] `@clerk/types`
- [ ] `@clerk/themes`
- [ ] `@clerk/localizations`
- [ ] `@clerk/clerk-expo`
- [ ] `@clerk/backend`
- [ ] `@clerk/clerk-sdk-node`
- [ ] `@clerk/shared`
- [ ] `@clerk/fastify`
- [ ] `@clerk/chrome-extension`
- [ ] `gatsby-plugin-clerk`
- [ ] `build/tooling/chore`

### CODE REVIEW

**Comment by LekoArts** [comment]
The `aria-label` items will be read out by a screen reader so they shouldn't be in snake case, dash case, etc. and be in a way that makes sense read out loud. If possible, don't use aria-label and use visible text or aria-describedby instead.

Everything you'd need to know about this is explained here, also with guides on what to use when :)
https://www.w3.org/WAI/ARIA/apg/practices/names-and-descriptions/

**Review by anagstef — APPROVED** [review]
Great job on this! 🎉

**Review by desiprisg — APPROVED** [review]

**Review by dimkl — APPROVED** [review]

**Review by LekoArts — CHANGES_REQUESTED** [review]

File: packages/clerk-js/src/ui/components/OrganizationSwitcher/OrganizationSwitcherTrigger.tsx [hunk 1]
```diff
@@ -33,6 +33,9 @@ export const OrganizationSwitcherTrigger = withAvatarShimmer(
         colorScheme='neutral'
         sx={[t => ({ minHeight: 0, padding: `0 ${t.space.$2} 0 0`, position: 'relative' }), sx]}
         ref={ref}
+        aria-label='organization-switcher-trigger'
```
Comment (LekoArts) [inline packages/clerk-js/src/ui/components/OrganizationSwitcher/OrganizationSwitcherTrigger.tsx:36]: This needs to be something more descriptive like this:

```suggestion
        aria-label={`${props.isOpen ? 'Close' : 'Open'} organization switcher`}
```

File: packages/clerk-js/src/ui/components/OrganizationSwitcher/OrganizationSwitcherPopover.tsx [hunk 2]
```diff
@@ -133,9 +133,14 @@ export const OrganizationSwitcherPopover = React.forwardRef<HTMLDivElement, Orga
...
           elementDescriptor={descriptors.organizationSwitcherPopoverCard}
           ref={ref}
+          role='dialog'
+          aria-label='organization switcher popover'
           {...rest}
         >
-          <PopoverCard.Main elementDescriptor={descriptors.organizationSwitcherPopoverMain}>
+          <PopoverCard.Main
+            elementDescriptor={descriptors.organizationSwitcherPopoverMain}
+            role='menu'
```
Comment (LekoArts) [inline packages/clerk-js/src/ui/components/OrganizationSwitcher/OrganizationSwitcherPopover.tsx:142]: This would only make sense if we'd use the rest of the menu syntax (https://developer.mozilla.org/en-US/docs/Web/Accessibility/ARIA/Roles/menu_role) which I think we don't. So in current state this should be dropped then

File: packages/clerk-js/src/ui/components/OrganizationSwitcher/OrganizationSwitcherPopover.tsx [hunk 3]
```diff
@@ -144,8 +149,9 @@ export const OrganizationSwitcherPopover = React.forwardRef<HTMLDivElement, Orga
                   organization={currentOrg}
                   user={user}
                   sx={theme => t => ({ padding: `0 ${theme.space.$6}`, marginBottom: t.space.$2 })}
+                  aria-hidden='true'
```
Comment (LekoArts) [inline packages/clerk-js/src/ui/components/OrganizationSwitcher/OrganizationSwitcherPopover.tsx:152]: Why do you hide this?

File: packages/clerk-js/src/ui/components/OrganizationSwitcher/OrganizationSwitcherTrigger.tsx [hunk 4]
```diff
@@ -33,6 +33,9 @@ export const OrganizationSwitcherTrigger = withAvatarShimmer(
...
+        aria-expanded={props.isOpen}
+        aria-haspopup='dialog'
```
Comment (LekoArts) [inline packages/clerk-js/src/ui/components/OrganizationSwitcher/OrganizationSwitcherTrigger.tsx:38]: The button will also need an `aria-controls` tied to the popover contents

File: packages/clerk-js/src/ui/components/OrganizationSwitcher/OrganizationSwitcherPopover.tsx [hunk 5]
```diff
@@ -144,8 +149,9 @@ export const OrganizationSwitcherPopover = React.forwardRef<HTMLDivElement, Orga
...
                 />
-                <Actions>
+                <Actions role='group'>
```
Comment (LekoArts) [inline packages/clerk-js/src/ui/components/OrganizationSwitcher/OrganizationSwitcherPopover.tsx:154]: Why did you add this here?

File: packages/clerk-js/src/ui/components/OrganizationSwitcher/OrganizationSwitcherPopover.tsx [hunk 6]
```diff
@@ -133,9 +133,14 @@ export const OrganizationSwitcherPopover = React.forwardRef<HTMLDivElement, Orga
         <PopoverCard.Root
           elementDescriptor={descriptors.organizationSwitcherPopoverCard}
           ref={ref}
+          role='dialog'
+          aria-label='organization switcher popover'
```
Comment (LekoArts) [inline packages/clerk-js/src/ui/components/OrganizationSwitcher/OrganizationSwitcherPopover.tsx:137]:
```suggestion
          aria-label='Organization switcher popover'
```

File: packages/clerk-js/src/ui/components/OrganizationSwitcher/OtherOrganizationActions.tsx [hunk 7]
```diff
@@ -44,7 +44,10 @@ export const OrganizationActionList = (props: OrganizationActionListProps) => {
   return (
     <>
       <UserInvitationSuggestionList />
-      <SecondaryActions elementDescriptor={descriptors.organizationSwitcherPopoverActions}>
+      <SecondaryActions
+        elementDescriptor={descriptors.organizationSwitcherPopoverActions}
+        role='group'
```
Comment (LekoArts) [inline packages/clerk-js/src/ui/components/OrganizationSwitcher/OtherOrganizationActions.tsx:49]: Same comment as above

File: packages/clerk-js/src/ui/components/UserButton/UserButtonTrigger.tsx [hunk 8]
```diff
@@ -20,6 +20,9 @@ export const UserButtonTrigger = withAvatarShimmer(
         variant='roundWrapper'
         sx={[theme => ({ borderRadius: theme.radii.$circle }), sx]}
         ref={ref}
+        aria-label='user-button-trigger'
```
Comment (LekoArts) [inline packages/clerk-js/src/ui/components/UserButton/UserButtonTrigger.tsx:23]: This will also need a better `aria-label` name.

Also an `aria-controls`

File: packages/clerk-js/src/ui/components/UserButton/UserButtonPopover.tsx [hunk 9]
```diff
@@ -69,9 +69,14 @@ export const UserButtonPopover = React.forwardRef<HTMLDivElement, UserButtonPopo
       <PopoverCard.Root
         elementDescriptor={descriptors.userButtonPopoverCard}
         ref={ref}
+        role='dialog'
```
Comment (LekoArts) [inline packages/clerk-js/src/ui/components/UserButton/UserButtonPopover.tsx:72]: Same as other popover

File: packages/clerk-js/src/ui/components/OrganizationSwitcher/UserMembershipList.tsx [hunk 10]
```diff
@@ -36,13 +36,16 @@ export const UserMembershipList = (props: UserMembershipListProps) => {
         overflowY: 'auto',
         ...common.unstyledScrollbar(t),
       })}
+      role='group'
+      aria-label={hidePersonal ? 'memberships' : 'accounts'}
```
Comment (LekoArts) [inline packages/clerk-js/src/ui/components/OrganizationSwitcher/UserMembershipList.tsx:40]: Needs to be more descriptive that this list is about

**Review by LekoArts — CHANGES_REQUESTED** [review]

**Comment by LekoArts** [comment]
https://codesandbox.io/s/distracted-swirles-jo1pvu?file=/src/Popover.tsx from https://floating-ui.com/docs/popover shows how to do popovers with our floating-ui usage

File: packages/clerk-js/src/ui/components/OrganizationSwitcher/OrganizationSwitcherTrigger.tsx (hunk 1 above)
Comment (panteliselef) [reply packages/clerk-js/src/ui/components/OrganizationSwitcher/OrganizationSwitcherTrigger.tsx:36]: @LekoArts I get why we want to remove  the `-`, but why do we need to add Close and Open ? Shouldn't we just rely on `aria-expanded` ?

File: packages/clerk-js/src/ui/components/OrganizationSwitcher/OrganizationSwitcherPopover.tsx (hunk 2 above)
Comment (panteliselef) [reply packages/clerk-js/src/ui/components/OrganizationSwitcher/OrganizationSwitcherPopover.tsx:142]: Nested components use the role of `menuitem`, do we need something more ?

File: packages/clerk-js/src/ui/components/OrganizationSwitcher/OrganizationSwitcherPopover.tsx (hunk 3 above)
Comment (panteliselef) [reply packages/clerk-js/src/ui/components/OrganizationSwitcher/OrganizationSwitcherPopover.tsx:152]: menu only allows for "interactive" elements to be menu items

File: packages/clerk-js/src/ui/components/OrganizationSwitcher/OrganizationSwitcherTrigger.tsx (hunk 4 above)
Comment (panteliselef) [reply packages/clerk-js/src/ui/components/OrganizationSwitcher/OrganizationSwitcherTrigger.tsx:38]: this is inside `...rest`

File: packages/clerk-js/src/ui/components/OrganizationSwitcher/OrganizationSwitcherPopover.tsx (hunk 5 above)
Comment (panteliselef) [reply packages/clerk-js/src/ui/components/OrganizationSwitcher/OrganizationSwitcherPopover.tsx:154]: Our code treats these actions and the `OrganizationActionList` as 2 different groups

i believe it is allowed to have groups inside a `menu`

File: packages/clerk-js/src/ui/components/UserButton/UserButtonTrigger.tsx (hunk 8 above)
Comment (panteliselef) [reply packages/clerk-js/src/ui/components/UserButton/UserButtonTrigger.tsx:23]: Same as above

File: packages/clerk-js/src/ui/components/OrganizationSwitcher/OrganizationSwitcherTrigger.tsx (hunk 1 above)
Comment (LekoArts) [reply packages/clerk-js/src/ui/components/OrganizationSwitcher/OrganizationSwitcherTrigger.tsx:36]: I can't link to W3C right now (it's down for me), so for ref the mozilla docs: https://developer.mozilla.org/en-US/docs/Web/Accessibility/ARIA/Attributes/aria-expanded#buttons
It's not necessary but a nice-to-have

I think in combination with the haspopup (https://www.accessibility-developer-guide.com/examples/sensible-aria-usage/expanded/#adding-haspopup-optional) this will be better

File: packages/clerk-js/src/ui/components/OrganizationSwitcher/OrganizationSwitcherTrigger.tsx (hunk 4 above)
Comment (LekoArts) [reply packages/clerk-js/src/ui/components/OrganizationSwitcher/OrganizationSwitcherTrigger.tsx:38]: When inspecting the current components with dev tools I didn't see this. Can you double check that those are present?

File: packages/clerk-js/src/ui/components/OrganizationSwitcher/OrganizationSwitcherPopover.tsx (hunk 3 above)
Comment (LekoArts) [reply packages/clerk-js/src/ui/components/OrganizationSwitcher/OrganizationSwitcherPopover.tsx:152]: aria-hidden should be used for these things:

- Purely decorative content, such as icons or images
- Duplicated content, such as repeated text
- Offscreen or collapsed content, such as menus

That's not the case for that item, thus we shouldn't hide it. You'll need to construct the menu in another way then

File: packages/clerk-js/src/ui/components/OrganizationSwitcher/OrganizationSwitcherTrigger.tsx (hunk 4 above)
Comment (panteliselef) [reply packages/clerk-js/src/ui/components/OrganizationSwitcher/OrganizationSwitcherTrigger.tsx:38]: It's the last one

image

Popover with the corresponding id
image

Comment (LekoArts) [reply packages/clerk-js/src/ui/components/OrganizationSwitcher/OrganizationSwitcherTrigger.tsx:38]: Ok 👍

File: .changeset/ten-wolves-cry.md
Comment (LekoArts) [inline .changeset/ten-wolves-cry.md:1]: Seems like a duplicate file here

File: .changeset/afraid-bobcats-mate.md [hunk 11]
```diff
@@ -0,0 +1,5 @@
+---
+'@clerk/clerk-js': patch
+---
+
+Impove accessibility of UserButton and OrganizationSwitcher by using aria-\* attributes and roles like `menu` and `menuitems`.
```
Comment (LekoArts) [inline .changeset/afraid-bobcats-mate.md:5]:
```suggestion
Improve accessibility of `<UserButton />` and `<OrganizationSwitcher />` by using `aria-*` attributes (where appropriate) and roles like `menu` and `menuitem`.
```

File: .changeset/ten-wolves-cry.md
Comment (panteliselef) [reply .changeset/ten-wolves-cry.md:1]: 😅 i will remove it

**Review by LekoArts — APPROVED** [review]
Thanks!

**Comment by clerk-cookie** [comment]
This PR has been automatically locked since there has not been any recent activity after it was closed. Please open a new issue for related bugs.
